pkg/metrics/metrics.go

```diff
@@ -55,6 +56,98 @@ const (
 	//L7DNS is the value used to report DNS label on metrics
 	L7DNS = "dns"
 )
+const (
+	// SubsystemBPF is the subsystem to scope metrics related to the bpf syscalls.
+	SubsystemBPF = "bpf"
+
+	// SubsystemDatapath is the subsystem to scope metrics related to management of
+	// the datapath. It is prepended to metric names and separated with a '_'.
+	SubsystemDatapath = "datapath"
+
+	// SubsystemAgent is the subsystem to scope metrics related to the cilium agent itself.
+	SubsystemAgent = "agent"
+
+	// SubsystemK8s is the subsystem to scope metrics related to Kubernetes
+	SubsystemK8s = "k8s"
+
+	// SubsystemK8sClient is the subsystem to scope metrics related to the kubernetes client.
+	SubsystemK8sClient = "k8s_client"
+
+	// SubsystemKVStore is the subsystem to scope metrics related to the kvstore.
+	SubsystemKVStore = "kvstore"
+
+	// SubsystemNodes is the subsystem to scope metrics related to the node manager.
+	SubsystemNodes = "nodes"
+
+	// SubsystemTriggers is the subsystem to scope metrics related to the trigger package.
+	SubsystemTriggers = "triggers"
+)
+
+const (
+	// SubsystemBPFMask is the mask used for BPF
+	SubsystemBPFMask uint64 = 1 << iota
+
+	// SubsystemDatapathMask is the mask used for Datapath
+	SubsystemDatapathMask
+
+	// SubsystemAgentMask is the mask used for agent
+	SubsystemAgentMask
+
+	// SubsystemK8sMask is the mask used for k8s
+	SubsystemK8sMask
+
+	// SubsystemK8sClientMask is the mask used for k8s client
+	SubsystemK8sClientMask
+
+	// SubsystemKVStoreMask is the mask used for kvstore
+	SubsystemKVStoreMask
+
+	// SubsystemNodesMask is the mask used for the node manager
+	SubsystemNodesMask
+
+	// SubsystemTriggersMask is the mask used in the pkg/trigger
+	SubsystemTriggersMask
+)
+
+// DefaultSubsystems returns the default subsystems that will be used to expose
+// metrics.
+func DefaultSubsystems() []string {
```

Comment:
I don't think the granularity of entire subsystems will be sufficient. Most subsystems will be scalable and less scalable metrics and users are likely to opt out of less scalable metrics.

I see two paths that make sense:

1. High-level flag that disables less scalable metrics such as histograms  
2. A per metric systems where we have a default `[]string{metric1,metric2,metric3}` and then allow to modify that with `--metrics=+metric2 --metrics=-metric3` to allow enabling or disabling individual metrics without requiring to be aware of the default all the time.

Reply:
@tgraf I hope that I'm not understanding 2) correctly, we will give the users to select over 50 different metrics, this sounds extremely overly complicated. If users don't want those metrics from being considered in prometheus they can configure prometheus to do that.

Reply:
How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable? I don't think we want users to get too involved. We want to have a good set of defaults and allow developers and advanced to enable individual metrics. We can also to `+grp-XXX` to enable individual groups.

I think subsystem is an arbitrary grouping that will be incredibly limiting if users want to find a good trade-off between visibility and load.

Reply:
>How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable?

Scalable where? In prometheus or Cilium? In Cilium the only non scalable metrics are the bpf related ones. In prometheus, users can specify which metrics they want to disable in its configuration.